include/lsst/partition/ObjectIndex.h

```diff
@@ -0,0 +1,195 @@
+ * no specific requirements for file locking or direct access are imposed by this implemenation.
+ *
+ * Notes on the thread-safety of the class's implementation:
+ * 1) All public methods of the class are thread safe.
+ * 2) An internal state of the index is guarded by a mutex.
+ */
+class ObjectIndex {
+public:
+    // Objects of this class can't be copied or directly constructed by clients.
+    ObjectIndex(ObjectIndex const&) = delete;
+    ObjectIndex& operator=(ObjectIndex const&) = delete;
+
+    /// The non-trivial destructor is needed to close and release resources used by the index.
+    ~ObjectIndex();
+
+    /// Modes for opening the index.
+    enum Mode {
+        READ,
+        WRITE
+    };
+
+    /**
+     * The factory method.
+     * 
+     * @return  A reference to the underlining implementation of the index
+     */
+    static ObjectIndex& instance() {
+        static ObjectIndex oi;
+        return oi;
+    }
+
+    /// @return A value of 'true' if the index is open or created.
+    bool isOpen() const { return _isOpen; }
+
+    /// @return A mode the index was open/created in.
+    Mode mode() const { return _mode; }
+
+    /**
+     * Open (or create) index for writing into a local file.
+     * 
+     * @param fileName  A path to a local file to be open/created.
+     * @param editor  A CSV "editor" which is used for formatting output products.
+     * @param idFieldName  The name of an output field representing object identifiers.
+     * @param chunkIdFieldName  The name of an output field representing chunk identifiers.
+     * @param subChunkIdFieldName  The name of an output field representing sub-chunk identifiers.
+     * @throw std::invalid_argument  Invalid specificaton of the fields.
+     * @throw std::runtime_error  If the file opening/creation operation failed for another reason.
+     */
+    void create(std::string const& fileName,
+                csv::Editor const& editor,
+                std::string const& idFieldName,
+                std::string const& chunkIdFieldName,
+                std::string const& subChunkIdFieldName);
+
+    /**
+     * Open the index.
+     * 
+     * Note that parameter 'dialect' should be normally the same one which is
+     * used for making output products of the partitioner. This is based on an assumption
+     * that all products made by the partitioner in a context of a given catalog would be
+     * partitoned in the same way. If that's not the case 
+     *
+     * @param url  An index specification.
+     * @param dialect  A dialect for parsing index specifications.
+     * @throw std::invalid_argument  Invalid specificaton of the index.
+     * @throw std::runtime_error  Failed to open the index.
+     */
+    void open(std::string const& url,
+              csv::Dialect const& dialect);
+
+    /// Close the index and release all resources associated with it. 
+    void close();
```

Comment:
As I explained it earlier in a reply to one of your comments, this class may get a more complex implementation with an interface having just two pure virtual methods `write` and `read`, and multiple implementations managing the lifecycle and the general state of the index.
I wouldn't do it right now since the road map of this development isn't clear to me (and, I guess, to anyone else). Let's just wait for the next "event" and do refactoring based on refined requirements to the index and its incarnations. My present focus is to make a tool suitable for KPM50 and some planned experimentations with existing catalogs (`objectId`-based partitioning of the "dependent" tables in `GAIA DR2`).